onnxruntime/core/providers/cpu/controlflow/if.cc

```diff
@@ -96,12 +96,23 @@ ONNX_CPU_OPERATOR_VERSIONED_KERNEL(If,
                                    If);
 
 // optional type is supported starting opset-16
+#if !defined(DISABLE_OPTIONAL_TYPE)
 ONNX_CPU_OPERATOR_KERNEL(If,
                          16,
                          KernelDefBuilder()
                              .TypeConstraint("B", DataTypeImpl::GetTensorType<bool>())
                              .TypeConstraint("V", DataTypeImpl::AllTensorAndSequenceTensorAndOptionalTypes()),
```

Comment:
>AllTensorAndSequenceTensorAndOptionalTypes [](http://example.com/codeflow?start=64&length=42)

Would it be simpler to just have AllTensorAndSequenceTensorAndOptionalTypes not return any optional types so that the operator registrations don't need updating? Just thinking about what happens when we inevitably make Sequence support optional as well.

Reply:
@edgchen1 is this going to be problematic with the kernel hashes either way? do we need the optional types to exist to some degree so the hash is consistent? otherwise you'd need to generate the model with exactly the same build as you run it which seems overly restrictive and confusing.

Reply:
as is, the hashes will be different given different types. not sure what the best way to handle it is.
registering the optional types here unconditionally and just not providing implementations for them in a minimal build might work. but then we need at least enough infrastructure to be able to specify optional types